File: camera.cpp
#include "camera.hpp"

#include <cmath>

namespace {
    double half_height(double const vfov) {
        double const theta = vfov * M_PI / 180;
        return std::tan(theta / 2);
    }
    
    double half_width(double const vfov, double const aspect) {
        return aspect * half_height(vfov);
    }

    vec3 w(vec3 const from, vec3 const at) {
        return (from - at).unit_vector();
    }

    vec3 u(vec3 const from, vec3 const at, vec3 const up) {
        return up.cross(w(from, at)).unit_vector();
    }

    vec3 v(vec3 const from, vec3 const at, vec3 const up) {
        return w(from, at).cross(u(from, at, up));
    }
}

camera::camera(vec3 const from, vec3 const at, vec3 const up, double const vfov, double const aspect)
    : origin{from},
      lower_left_corner{from - half_width(vfov, aspect) * u(from, at, up) - half_height(vfov) * v(from, at, up) - w(from, at)},
      horizontal{2 * half_width(vfov, aspect) * u(from, at, up)},
      vertical{2 * half_height(vfov) * v(from, at, up)} {}

ray camera::get_ray(double const u, double const v) const {
    return {
        origin,
        // This is *not* a unit vector which apparently allows for simpler/faster code?
        lower_left_corner + u * horizontal + v * vertical - origin
    };
}

File: rrt/src/camera.rs
use crate::pt3::Pt3;
use crate::ray::Ray;
use crate::vec3::Vec3;

pub trait Camera {
    fn get_ray(&self, u: f32, v: f32) -> Ray;
}

#[derive(Debug)]
pub struct SimpleCamera {
    origin: Pt3,
    lower_left_corner: Pt3,
    horizontal: Vec3,
    vertical: Vec3,
}

impl SimpleCamera {
    // vfov is top to bottom in degrees
    pub fn fov(look_from: Pt3, look_at: Pt3, vup: Vec3, vfov: f32, aspect: f32) -> Self {
        let theta = vfov * std::f32::consts::PI / 180.0;
        let half_height = (theta / 2.0).tan();
        let half_width = aspect * half_height;

        let w = (look_from - look_at).unit_vector();
        let u = vup.cross(w).unit_vector();
        let v = w.cross(u);

        SimpleCamera {
            origin: look_from,
            lower_left_corner: look_from - (half_width * u) - (half_height * v) - w,
            horizontal: 2.0 * half_width * u,
            vertical: 2.0 * half_height * v,
        }
    }
}

impl Camera for SimpleCamera {
    fn get_ray(&self, u: f32, v: f32) -> Ray {
        let direction = self.lower_left_corner + u * self.horizontal + v * self.vertical;
        Ray {
            origin: self.origin,
            direction: direction - self.origin,
        }
    }
}
